### PR TITLE
WIP Support Firebase JS SDK 5.1 and Firestore with v3

### PR DESCRIPTION
**Pull request by jamesdaniels**
### Description

**This is a work in progress!** A lot has changed since version 2 of emberfire was released. We wanted our third major release to work great with Ember v3+, Fastboot, Cloud Firestore, Cloud Functions, and modern security rules out of the box. We want to also deal completely with the weak points of version 2; namely relationships, querying, and performance.

Given all that has changed and all we want to accomplish, we thought best to rewrite.

I'll write up some more as this release closes but I wanted to throw my work in progress up here for people to check out. The work here is loosely based on my v3 proposal here #542 (I've refined a couple details while taking feedback and building things out).

Related issues #546 , #552 , #531, #529, #522, #512, #511, #465, #461, #400, etc.

To try out this branch run `ember install emberfire-exp`.

#### Cloud Firestore adapter with offline-persistence enabled

```js
// app/adapters/application.js
import FirestoreAdapter from 'emberfire/adapters/firestore';

export default FirestoreAdapter.extend({
    enablePersistence: true,
    persistenceSettings: { experimentalTabSynchronization: true }
})
```

#### Firestore backed model with embedded records, subcollections, and queries

```js
// app/models/post.js
import DS from 'ember-data';
const { belongsTo, hasMany } = DS;

export default DS.Model.extend({
    user: belongsTo('user'),
    tags: hasMany('tags', { embedded: true }),
    comments: hasMany('comments', { subcollection: true, query: ref => ref.orderBy('publishedAt', 'asc') })
});
```

#### Route demonstrating Firestore querying capabilities & realtime route mixin

```js
// app/routes/posts.js
import Route from '@ember/routing/route';
import { RealtimeRouteMixin } from 'emberfire/services/realtime-listener';

export default Route.extend(RealtimeRouteMixin, {
    model(params) {
        return this.store.query('post', ref => {
            const retVal = ref.orderBy('publishedAt', 'desc');
            if (params.tag) { return retValue.where(`tags.${params.tag}`, '==', true) }
            return retVal;
        })
    }
});
```

#### Firebase app service and `ember-simple-auth` integration

```js
// app/controllers/application.js
import Controller from '@ember/controller';
import { inject as service } from '@ember/service';
import { get } from '@ember/object';
import firebase from 'firebase/app';

export default Controller.extend({
    session: service(),
    firebaseApp: service(),
    actions: {
        logout() {
            return get(this, 'session').invalidate();
        },
        async login() {
            const provider = new firebase.auth.GoogleAuthProvider();
            const auth = await get(this, 'firebaseApp').auth();
            return auth.signInWithPopup(provider);
        }
    }
});
```

#### Support for multiple Firebase applications / services

```js
// config/environment.js

module.exports = function(environment) {
  let ENV = {
    firebase: [
      {
        apiKey: "...",
        authDomain: "YOUR_APP.firebaseapp.com",
        databaseURL: "https://YOUR_APPfirebaseio.com",
        projectId: "YOUR_APP",
        storageBucket: "YOUR_APP.appspot.com",
        messagingSenderId: "00000000"
      }, {
        apiKey: "...",
        authDomain: "YOUR_SECOND_APP.firebaseapp.com",
        databaseURL: "https://YOUR_SECOND_APP.firebaseio.com",
        projectId: "YOUR_SECOND_APP",
        storageBucket: "YOUR_SECOND_APP.appspot.com",
        messagingSenderId: "0000000",
        name: "second-app"
      }
    ],
    ...
```

After initialization there will now be a `service:firebase-app` and `service:firebase-second-app` ready for injection.

### TODO

- [x] Get the first pass of the realtime service in
- [x] Get the test suite together and passing
  - [ ] Flush out the test suite
  - [x] Give the relationship system a good once over
- [ ] Setup Travis with the NPM keys and cut beta.1
- [x] Consider Fastboot Authentication (admin cookies, express middleware?)
- [ ] Put together migration guide
- [ ] Update the documentation
- [x] Publish API reference
  - [ ] Add class descriptions, parameter comments, code examples, and more tsdocs
  - [ ] Add links to the API reference from the docs
  - [ ] Clean up external types
  - [ ] Remove external `Defined in` references to my node_modules
- [x] Reassess `ember-auto-import`, `browserify`, and `browserify-shim` as we near final
- [x] Reassess shoebox as we near final
- [ ] Initializer should bring in the serializer
- [ ] Should not require ember-simple-auth, etc.

*more to come...*

### CODE REVIEW

**Comment by jamesdaniels** [comment]
@vabtrix yes, authentication will be done via the Firebase SDK's API. Built in providers for torii and ember-simple-auth will pick it up automatically, or you could build auth state management yourself.

**Comment by aalimovs** [comment]
Would this be merged after Firestore hits GA?

**Comment by arunartoo** [comment]
Any update on when we can get a beta version of the official emberfire with Firestore support? We have been waiting for months... :( 
Looks like ember is left behind.

**Comment by jamesdaniels** [comment]
I'm aiming to release AngularFire v5 this week and then will be back on Emberfire v3. I had to give that lib my priority since it has 30x the downloads. 

In doing that work I also shook out some bugs with the JS SDK that would have also affected Ember, so the work isn't happening in a vacuum.

**Comment by lukaiser** [comment]
@jamesdaniels any news? It has been a other 7 weeks and no commits since June. It gets hard to sell to my boss to keep waiting and not switch to an other backend. Is there any way to persuade google to allocate more manpower to it?

**Comment by brancusi** [comment]
@jamesdaniels, I'm sure you are spread very thin across all the projects, I don't mean to nag, but any estimates would be much appreciated. We have a new project in the works and the team just wants to get a sense if this project will continue to be supported before we commit to FB again. Thank you

**Comment by oskarrough** [comment]
Hi all. I believe now it is clear that many, me included, are excited for the next Emberfire release. So let's try to organize and help each other actually publish it. Go through the todos above in this PR, go through the https://github.com/firebase/emberfire/blob/v3/docs/quickstart.md, try to create something and report back here with feedback or send a PR.

### Installation

FIrst off, how do you properly install this branch to test?

Tried `yarn add https://github.com/firebase/emberfire.git#v3` but addon won't install correctly.

This seems to work

```
git clone https://github.com/firebase/emberfire.git
cd emberfire
git checkout v3
yarn
yarn run ts:precompile
yarn link
```

…and:

```
ember new my-app
cd my-app
yarn link emberfire
```

@jamesdaniels, is this the right way to test it? Would it make sense to simplify with a prerelease on NPM? 

### Blueprints

Now we need to run the default blueprint. I tried `ember g emberfire` which fails.

```shell
$ ember g emberfire
installing emberfire
The `ember generate <entity-name>` command requires an entity name to be specified. For more details, use `ember help`.
```

Apparently we have to namespace them, but the blueprint seems to fail because of the symlink. We can update the guides here.

``` shell
$ ember g emberfire emberfire
installing emberfire
  identical app/adapters/application.js
  install package firebase
  install package emberfire
  install addon ember-auto-import
Command failed: yarn add --dev emberfire@^3.0.0 --non-interactive
error Couldn't find any versions for "emberfire" that matches "^3.0.0"
``` 

I installed them manually to proceed.

Next and finally, I see this when the app boots and now I'm stuck.

```
jQuery.Deferred exception: Class constructor FirestoreAdapter cannot be invoked without 'new' TypeError: Class constructor FirestoreAdapter cannot be invoked without 'new'
    at new Class (http://localhost:4200/assets/vendor.js:50903:72)
    at Function.create (http://localhost:4200/assets/vendor.js:50916:16)
    at FactoryManager.create (http://localhost:4200/assets/vendor.js:28662:39)
    at instantiateFactory (http://localhost:4200/assets/vendor.js:28511:35)
    at _lookup (http://localhost:4200/assets/vendor.js:28454:16)
    at Container.lookup (http://localhost:4200/assets/vendor.js:28353:20)
    at Class.lookup (http://localhost:4200/assets/vendor.js:49364:33)
    at Class.adapterFor (http://localhost:4200/assets/vendor.js:75388:52)
    at Class._flushPendingFetchForType (http://localhost:4200/assets/vendor.js:73562:27)
    at Map.forEach (<anonymous>) undefined
```

@runspired, is this what you were seeing, too? Some babel + `new Class` stuff.

Would also be great to organize and clarify the todos in your PR so we have a better chance of contributing. To the batmobile!

**Comment by jamesdaniels** [comment]
I'll need to hunt down the right person to allow me to cut prereleases to NPM, or do it under a personal scope, as our current build system doesn't allow me to provide a tag. At the very least I can compile and provide a tarball here, rather than having folks compile it themselves.

Thanks for raising the install and initial setup issues @oskarrough, that's very helpful. I'll look into those issues and see if I can get them addressed so you can get up and running with this.

**Comment by jamesdaniels** [comment]
Also I've spoken to folk and I'm going to err on the side of cutting v3 early, as it's clear that we don't have the capacity to both work on v3 and maintain v2. Have to try not to let perfectionism get in the way of "good enough".

**Comment by runspired** [comment]
@oskarrough 

I ended up rolling my own firebase adapter and auth bc I couldn't get the package to work with more recent versions of ember and ember-cli and because I both wanted to use firestore and subscribe to the records queried.

**Comment by arun-prasad-r** [comment]
@jamesdaniels any updates on v3 release or a prerelease with firestore support? 

Also, can you please provide some steps that we need to follow to properly set up and test this branch locally?

**Comment by arun-prasad-r** [comment]
@jamesdaniels any news on the release? We like firebase and want to use cloud firestore and we don't want to switch to Angular just because firestore support is not available for ember. We've invested a significant amount of time in learning the framework and its ecosystem.

**Comment by brendanoh** [comment]
Came here to start a new EmberFire App and am pretty confused on whether EmberFire is even supported anymore.  

@runspired any chance you could release (or gist) your "firebase adapter and auth" as I am sure a lot of us would be appreciative.

**Comment by chadRoberge** [comment]
Any update or timeframe for this?

File: package.json [hunk 1]
```diff
@@ -1,98 +1,79 @@
...
+    "ember-cli-typescript": "^1.3.2",
+    "ember-cli-uglify": "^2.1.0",
+    "ember-data": "~3.1.1",
+    "ember-disable-prototype-extensions": "^1.1.3",
+    "ember-export-application-global": "^2.0.0",
+    "ember-fetch": "^5.0.0",
+    "ember-load-initializers": "^1.1.0",
+    "ember-resolver": "^4.5.5",
+    "ember-simple-auth": "^1.6.0",
+    "ember-source": "~3.2.0",
```
Comment (zoltan-nz) [inline package.json]: Thank you! Great to see that our favourite library is moving forward. Great moment. :) I'm looking forward to testing with our popular [Ember.js tutorial](https://yoember.com) app.

If things are going well maybe we could try out to upgrade the addon to Ember 3.4 LTS or maybe Ember 3.7 Latest? What do you think?

**Comment by jamesdaniels** [comment]
I've published on NPM under `emberfire-exp` for the time being: `npm i --save-dev emberfire-exp`

**Comment by zoltan-nz** [comment]
After a quick test:

1. I noticed, that `ember-simple-auth` cannot be optional, it has be installed to the project with `ember install ember-simple-auth`. (v1.8.1 is installed)
2. The default adapter configuration changed and has to be updated to `FirestoreAdapter` as you already mentioned in the new guide

```
import FirestoreAdapter from 'emberfire/adapters/firestore';
export default FirestoreAdapter.extend({
});
```

3. After my demo app compiled and opened, however when the router wanted to access to the store/adapter first time, the following error appeared:

```
Class constructor FirestoreAdapter cannot be invoked without 'new' TypeError: Class constructor FirestoreAdapter cannot be invoked without 'new'
```

My app uses Ember / EmberData v3.4 LTS.

**Comment by jamesdaniels** [comment]
Yeah I'll make simple auth an optional dependency. Also I noticed that the serializer isn't coming in, so you'll want to create an application serializer that extends the one in Emberfire. I'll fix that shortly. 

Not sure about your 'new' issue. Might be the ember version.

File: package.json (hunk 1 above)
Comment (jamesdaniels) [reply package.json]: I'll bump to @latest here and rely on .ember-try for LTS.

**Comment by samselikoff** [comment]
@zoltan-nz I ran into the same error you did:

> Class constructor FirestoreAdapter cannot be invoked without 'new' TypeError: Class constructor FirestoreAdapter cannot be invoked without 'new'

Did you ever figure it out?

**Comment by zoltan-nz** [comment]
@samselikoff do you use Ember v3.4? I think, the latest beta version of EmberFire works fine with Ember v3.5+

So the solution was to upgrade Ember.

I would say, we should use EmberFire v2 until Ember v3.4 and use EmbeFire v3(beta) with Ember v3.5+

**Comment by sergchernata** [comment]
Experiencing this issue on:
```
"ember-source": "~3.6.0",
"ember-data": "~3.6.0",
"emberfire": "^3.0.0-rc.3",
```

**Pull request merged**